source/extensions/filters/network/connection_limit/connection_limit.h

```diff
@@ -0,0 +1,106 @@
+
+#include "common/common/thread_synchronizer.h"
+#include "common/runtime/runtime_protos.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ConnectionLimitFilter {
+
+/**
+ * All connection limit stats. @see stats_macros.h
+ */
+#define ALL_CONNECTION_LIMIT_STATS(COUNTER, GAUGE)                                                 \
+  COUNTER(connection_limited)                                                                      \
+  GAUGE(active_connections, Accumulate)
+
+/**
+ * Struct definition for connection limit stats. @see stats_macros.h
+ */
+struct ConnectionLimitStats {
+  ALL_CONNECTION_LIMIT_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+};
+
+/**
+ * Configuration shared across all connections. Must be thread safe.
+ */
+class Config : Logger::Loggable<Logger::Id::filter> {
+public:
+  Config(const envoy::extensions::filters::network::connection_limit::v3::ConnectionLimit&
+             proto_config,
+         Event::Dispatcher& dispatcher, Stats::Scope& scope, Runtime::Loader& runtime);
+
+  bool incrementConnectionWithinLimit();
+  void incrementConnection();
+  bool decrementConnection();
+  bool enabled() { return enabled_.enabled(); }
+  absl::optional<std::chrono::milliseconds> delay() { return delay_; }
+  ConnectionLimitStats& stats() { return stats_; }
+
+private:
+  static ConnectionLimitStats generateStats(const std::string& prefix, Stats::Scope& scope);
+  void onDelayTimer();
+
+  Runtime::FeatureFlag enabled_;
+  ConnectionLimitStats stats_;
+  const uint64_t max_connections_;
+  std::atomic<uint64_t> connections_;
+  absl::optional<std::chrono::milliseconds> delay_;
+  const Event::TimerPtr delay_timer_;
+
+  friend class ConnectionLimitTestBase;
+};
+
+using ConfigSharedPtr = std::shared_ptr<Config>;
+
+/**
+ * Per-connection connection limit filter
+ */
+class Filter : public Network::ReadFilter,
+               public Network::ConnectionCallbacks,
+               Logger::Loggable<Logger::Id::filter> {
+public:
+  Filter(const ConfigSharedPtr& config) : config_(config), is_rejected_(false) {}
+
+  // Network::ReadFilter
+  Network::FilterStatus onData(Buffer::Instance&, bool) override {
+    if (is_rejected_.load()) {
+      return Network::FilterStatus::StopIteration;
+    }
+    return Network::FilterStatus::Continue;
+  }
+
+  Network::FilterStatus onNewConnection() override;
+
+  void initializeReadFilterCallbacks(Network::ReadFilterCallbacks& read_callbacks) override {
+    read_callbacks_ = &read_callbacks;
+    read_callbacks_->connection().addConnectionCallbacks(*this);
+  }
+
+  // Network::ConnectionCallbacks
+  void onEvent(Network::ConnectionEvent event) override;
+  void onAboveWriteBufferHighWatermark() override {}
+  void onBelowWriteBufferLowWatermark() override {}
+
+private:
+  const ConfigSharedPtr config_;
+  Network::ReadFilterCallbacks* read_callbacks_{};
+  Event::TimerPtr delay_timer_;
+  std::atomic<bool> is_rejected_;
```

Comment:
I am not sure about this.
I want to make sure the onData() will StopIteration for the connections within the delay closing duration.
So we need a flag to indicate,  if one Filter is only processed by one thread, then I think it does not need to be atomic.

Reply:
A `Filter` lives inside a worker thread and should not be accessed from other threads.

Reply:
Sounds good, I will use bool instead of std::atomic<bool>.

Reply:
Thanks @rojkov . What about timer callbacks? Are they guaranteed to happen on the same worker thread?

Reply:
It's worth checking. But I expect timer callbacks to be events placed to a worker's event loop.

Reply:
Just realized this object is not even touched inside timer callback, so that shouldn't be concern. Also since the timer is created on the connection's dispatcher I guess it must fire on the same worker thread. So @rojkov's suggestions sounds right that we don't need the atomic object.

Reply:
Did a quick test, the timer should be triggered on the same work thread, changed to `atomic<boo>` to `bool`.